### PR TITLE
ministry names endpoint added, a custom select component added

### PR DESCRIPTION
**Pull request by iman-jamali-fw**
## Summary
This PR brings dynamic ministry names in a form of a select component for form designer. 
Implements #484 

## Changes
- An endpoint fetching ministry names from ODS was added to `webapi`
- That endpoint caches the results for an hour, to prevent excess calls to ODS
- A new `Select` component was added, extending formio `Select` component to support authenticated calls to the `webapi` backend.

## Screenshots (if applicable)
Screen Shot 2022-08-18 at 3 18 02 PM

### CODE REVIEW

File: forms-flow-web/src/formComponents/RemoteSelect/index.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    // Add filter capability
+    if (this.component.filter) {
+      url += (!url.includes('?') ? '?' : '&') + this.interpolate(this.component.filter);
+    }
+
+    // Set ignoreCache if it is
+    options.ignoreCache = this.component.ignoreCache;
+
+    // Set the required headers for connecting to the form-flow-weapi endpoints
+    headers = setFormFlowApiHeaders(headers);
```
Comment (iman-jamali-fw) [inline forms-flow-web/src/formComponents/RemoteSelect/index.js:145]: This was added

File: forms-flow-web/src/formComponents/RemoteSelect/index.js [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+          this.itemsLoadedResolve();
+          return;
+        }
+        let { url } = this.component.data;
+        let method;
+        let body;
+
+        if (url.startsWith('/')) {
+          // Adding the forms-flow-webapi base URL to the URL         
+          url = WEB_BASE_URL + url;
```
Comment (iman-jamali-fw) [inline forms-flow-web/src/formComponents/RemoteSelect/index.js:226]: This was added

File: forms-flow-web/src/formComponents/RemoteSelect/index.js [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+              this.setItems(items);
+            });
+          };
+        }
+      }
+    }
+  }
+}
+
+const setFormFlowApiHeaders = (headers) => {
```
Comment (iman-jamali-fw) [inline forms-flow-web/src/formComponents/RemoteSelect/index.js:299]: And this function was added

File: forms-flow-web/src/formComponents/RemoteSelect/index.js [hunk 4]
```diff
@@ -0,0 +1,303 @@
+import { Components, Formio } from 'react-formio';
+import _ from 'lodash';
+import NativePromise from 'native-promise-only';
+import { WEB_BASE_URL } from "../../apiManager/endpoints/config";
+import UserService from "../../services/UserService";
+
+const SelectComponent = Components.components.select;
+
+export default class RemoteSelect extends SelectComponent {
```
Comment (simensma-fresh) [inline forms-flow-web/src/formComponents/RemoteSelect/index.js:9]: I'm wondering if we could add a "MinistrySelect" component that extends this that defaults the url to the ministry name endpoint, just to make it as easy to use as possible for a forms designer

File: docker-compose.yml [hunk 5]
```diff
@@ -256,7 +256,8 @@ services:
       KEYCLOAK_ADMIN_PASSWORD: ${KEYCLOAK_ADMIN_PASSWORD}
       WEB_API_BASE_URL: ${FORMSFLOW_API_URL}
       EMPLOYEE_DATA_API_URL: ${EMPLOYEE_DATA_API_URL}
-      EMPLOYEE_DATA_AUTH_TOKEN: ${EMPLOYEE_DATA_AUTH_TOKEN}
+      ODS_AUTH_TOKEN: ${ODS_AUTH_TOKEN}
```
Comment (simensma-fresh) [inline docker-compose.yml:259]: The `EMPLOYEE_DATA_AUTH_TOKEN` is also used in the Openshift templates for camunda and the API, so you'll have to update those as well, or revert the change

**Review by simensma-fresh — CHANGES_REQUESTED** [review]

**Review by david-sabadash — APPROVED** [review]

File: docker-compose.yml (hunk 5 above)
Comment (iman-jamali-fw) [reply docker-compose.yml:259]: I thought `ODS_AUTH_TOKEN` is more general, now that we use it for more than one endpoint. Will update the Openshift once deployed

File: forms-flow-web/src/formComponents/RemoteSelect/index.js (hunk 4 above)
Comment (iman-jamali-fw) [reply forms-flow-web/src/formComponents/RemoteSelect/index.js:9]: Sure that works. I'll look into it, amongst other ways to make it easier to use for a designer.

Comment (iman-jamali-fw) [reply forms-flow-web/src/formComponents/RemoteSelect/index.js:9]: Addressed. A new "MinistrySelect" was added that gets values from the `/ministry-names` endpoint by default. I also made some adjustments to `RemoteSelect` to only support URL-based values by default.

File: forms-flow-web/src/formComponents/MinistrySelect/index.js [hunk 6]
```diff
@@ -0,0 +1,210 @@
+import { Components, Formio } from 'react-formio';
+import _ from 'lodash';
+import { WEB_BASE_URL } from "../../apiManager/endpoints/config";
+import UserService from "../../services/UserService";
+
+const SelectComponent = Components.components.select;
+
+export default class MinistrySelect extends SelectComponent {
```
Comment (simensma-fresh) [inline forms-flow-web/src/formComponents/MinistrySelect/index.js:8]: I would suggest extending `RemoteSelect` here instead of `SelectComponent` so we don't have to duplicate the loadItems / updateItems methods. 

```javascript
export default class MinistrySelect extends RemoteSelect {
      static schema(...extend) {
       ...
      }
      static get builderInfo() {
              ...
      }
}
```

**Review by simensma-fresh — CHANGES_REQUESTED** [review]

File: docker-compose.yml (hunk 5 above)
Comment (simensma-fresh) [reply docker-compose.yml:259]: For sure! The templates are defined in the `deployment/openshift` folder and are automatically applied when we merge things, so you'll have to update `camunda_dc.yaml` and `webapi_dc.yaml` before we can merge this 🙃

**Review by simensma-fresh — APPROVED** [review]

**Review by simensma-fresh — CHANGES_REQUESTED** [review]
REalized we have to add `MINISTRY_NAMES_API_URL` to the webapp Deployment config as well 🙃

**Review by simensma-fresh — APPROVED** [review]

**Review by simensma-fresh — APPROVED** [review]
